sklearn/cluster/_kmeans.py

```diff
@@ -778,6 +791,16 @@ class _BaseKMeans(
 ):
     """Base class for KMeans and MiniBatchKMeans"""
 
+    _parameter_constraints = {
+        "n_clusters": [Interval(Integral, 1, None, closed="left")],
+        "init": [StrOptions({"k-means++", "random"}), callable, "array-like"],
+        "n_init": [Interval(Integral, 1, None, closed="left")],
+        "max_iter": [Interval(Integral, 1, None, closed="left")],
+        "tol": [Interval(Real, 0, None, closed="left")],
+        "verbose": [Interval(Integral, 0, None, closed="left"), bool],
+        **get_random_state_param_constraints(),
```

Comment:
Not sure about this notation. The dict is `param_name : list of constraints` and the parameter is valid if it satisfies one of the constraints. For random_state it's `[None, RandomState, Interval(...)]`. I made this util to avoid typing the same thing everywhere because random_state is a widely used param.

Formally, the list [a, b, c] is the constraint `a or b or c`. I made the language abuse to name it a list of constraints. It's not perfect but it makes the wording kind of intuitive.

Reply:
I agree with @thomasjpfan here. We can have certain "reserved values" like we have for `"array-like"` and `"random_state"` could be another one.